File: source/java/nl/esciencecenter/xenon/examples/filesystems/DownloadFileSftpToLocalAbsolutePaths.java
.. code-tab:: java

    import nl.esciencecenter.xenon.credentials.PasswordCredential;
    import nl.esciencecenter.xenon.filesystems.CopyMode;
    import nl.esciencecenter.xenon.filesystems.CopyStatus;
    import nl.esciencecenter.xenon.filesystems.FileSystem;
    import nl.esciencecenter.xenon.filesystems.Path;

    public class DownloadFileSftpToLocalAbsolutePaths {

        public static void main(String[] args) throws Exception {

            // use the sftp file system adaptor to create a file system representation; the remote
            // filesystem requires credentials to log in, so we'll have to create those too.
            String adaptorRemote = "sftp";
            String location = "localhost:10022";
            String username = "xenon";
            char[] password = "javagat".toCharArray();
            PasswordCredential credential = new PasswordCredential(username, password);
            FileSystem filesystemRemote = FileSystem.create(adaptorRemote, location, credential);

            // define which file to download
            Path fileRemote = new Path("/home/xenon/sleep.stdout.txt");

            // use the local file system adaptor to create a file system representation
            String adaptorLocal = "file";
            FileSystem filesystemLocal = FileSystem.create(adaptorLocal);

            // define what file to download to
            Path fileLocal = new Path("/home/tutorial/xenon/sleep.stdout.txt");

            // create the destination file only if the destination path doesn't exist yet
            CopyMode mode = CopyMode.CREATE;

            // no need to recurse, we're just downloading a file
            boolean recursive = false;

            // perform the copy/download and wait 1000 ms for the successful or otherwise
            // completion of the operation
            String copyId = filesystemRemote.copy(fileRemote, filesystemLocal, fileLocal, mode, recursive);
            long timeoutMilliSecs = 1000;
            CopyStatus copyStatus = filesystemRemote.waitUntilDone(copyId, timeoutMilliSecs);

            // print any exceptions
            if (copyStatus.getException() != null) {
                System.out.println(copyStatus.getException().getMessage());
            } else {
                System.out.println("Done");
            }
        }
    }

File: source/python/examples/filesystems/download_file_sftp_to_local_absolute_paths.py
.. code-tab:: python

    from xenon import Server, PasswordCredential, CopyMode, Path

    with Server() as xenon:
        # use the sftp file system adaptor to create a file system
        # representation; the remote filesystem requires credentials to log in,
        # so we'll have to create those too.
        credential = PasswordCredential(
                username='xenon',
                password='javagat')

        remote_fs = xenon.create_file_system(
                adaptor='sftp',
                location='localhost:10022',
                credential=credential)

        # use the local file system adaptor to create a file system
        # representation
        local_fs = xenon.create_file_system(
                adaptor='file')

        # define which file to download
        remote_file = Path('/home/xenon/sleep.stdout.txt')

        # define what file to download to
        local_dir = Path('/home/tutorial/xenon/')
        local_file = local_dir.joinpath(remote_file.name)

        # create the destination file only if the destination path doesn't
        # exist yet
        mode = CopyMode.CREATE

        # no need to recurse, we're just downloading a file
        recursive = False

        # perform the copy/download and wait 1000 ms for the successful or
        # otherwise completion of the operation
        copy_id = remote_fs.copy(
                remote_file, local_fs, local_file,
                mode=mode, recursive=recursive)
        copy_status = remote_fs.wait_until_done(copy_id, timeout=1000)

        if copy_status.error:
            print(copy_status.error)
        else:
            print('Done')

        # remember to close the FileSystem instances, or use them as
        # context managers
        remote_fs.close()
        local_fs.close()
